### DISCUSSION TITLE
Understanding `parallelism` and `namespaceParallelism` in workflow-controller

### DISCUSSION THREAD

**Discussion by vadasambar** [argoproj/argo-workflows#12535] [Q&A] [answered]
https://github.com/argoproj/argo-workflows/blob/1202ae473a9f047cccef820a80ecb97c59a02b92/docs/workflow-controller-configmap.yaml#L22-L30 

Both `parallelism` and `namespaceParallelism` refer to `workflows that can execute`. 

Looking at the code (check below) it seems like the limit refers to the number of workflows that would be processed by the workflow-controller. It doesn't necessarily mean `number of workflows running` at any given point in time (it may or may not be the case). 

Here's where I am looking in the code (I am looking at 3.4.10 version of the workflow-controller):
* admit check: https://github.com/argoproj/argo-workflows/blob/1984c1ae47a126440076653c660e521a9c548074/workflow/controller/controller.go#L743-L750
* https://github.com/argoproj/argo-workflows/blob/1984c1ae47a126440076653c660e521a9c548074/workflow/controller/controller.go#L204-L207
* https://github.com/argoproj/argo-workflows/blob/a9ea0a5414d77d98693c6a80219285eab46697d1/workflow/sync/throttler.go#L135-L138
* wf added to the throttler:
  * https://github.com/argoproj/argo-workflows/blob/1984c1ae47a126440076653c660e521a9c548074/workflow/controller/controller.go#L850
  * https://github.com/argoproj/argo-workflows/blob/1984c1ae47a126440076653c660e521a9c548074/workflow/controller/controller.go#L863
 
1. If my understanding is correct, is there anyway to ensure there would be 100 `Running` workflows at any given point in time if say 500 workflows objects have been created? `parallelism` and `namespaceParallelism` ensure max workflows are at say 100 (if that's the limit I set) but there doesn't seem to a way to ensure at-least 100 workflows are `Running` at any given point in time. 

P.S.: 
2. Does this also mean that if the controller has already processed 100 workflows (`parallelism` being set at 100) and all of them are in `Running` state, throttler would allow next 100 batch of workflows and thereby making the total number of `Running` workflows > 100?

**Comment by vadasambar** [comment]
If this has already been talked somewhere, feel free to point me to that 🙏 (I couldn't find any issue or question like this)

**Comment by agilgur5** [comment, marked as answer]
> 1. If my understanding is correct, is there anyway to ensure there would be 100 `Running` workflows at any given point in time if say 500 workflows objects have been created? `parallelism` and `namespaceParallelism` ensure max workflows are at say 100 (if that's the limit I set) but there doesn't seem to a way to ensure at-least 100 workflows are `Running` at any given point in time.

Yes, this is correct. It ensures there are no more than 100 "incomplete" Workflows, not that there are 100 `Running` Workflows

> 2. Does this also mean that if the controller has already processed 100 workflows (`parallelism` being set at 100) and all of them are in `Running` state, throttler would allow next 100 batch of workflows and thereby making the total number of `Running` workflows > 100?

No, Workflows are only `remove`d from the `throttler` [when they are completed](https://github.com/argoproj/argo-workflows/blob/1984c1ae47a126440076653c660e521a9c548074/workflow/controller/controller.go#L756). The `wfQueue` responsible for processing and the `throttler` responsible for limiting are two separate data structures.

**Reply by vadasambar** [reply]
Thank you @agilgur5.
